File: arete-hash/src/fixture.rs
use std::collections::HashSet;

use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};

use crate::{
    canonicalize_jcs, hash_jcs, parse_json_bytes_strict, DecoderFixtureSet, HashError, HashId,
    IdlNormalized,
};

pub const DECODER_FIXTURE_SCHEMA_V1: &str = "arete.decoder-fixtures/v1";
pub const DECODER_FIXTURE_PUBLIC_VALUE_DIGEST_PREFIX: &str = "sha256:";
pub const DECODER_FIXTURE_MAX_CASES: usize = 256;
pub const DECODER_FIXTURE_MAX_ACCOUNT_BYTES: usize = 1024 * 1024;
pub const DECODER_FIXTURE_MAX_TOTAL_ACCOUNT_BYTES: usize = 8 * 1024 * 1024;
pub const DECODER_FIXTURE_ACCOUNT_DECODE_ERROR_CATEGORIES: &[&str] = &[
    "owner_mismatch",
    "unknown_account_type",
    "account_type_mismatch",
    "ambiguous_account_type",
    "account_decode_failed",
];

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct DecoderFixtureSetV1 {
    pub schema: String,
    pub program_id: String,
    pub normalized_idl_hash: HashId<IdlNormalized>,
    pub decoder_engine_id: String,
    pub decoder_abi_version: String,
    pub cases: Vec<DecoderFixtureCaseV1>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct DecoderFixtureCaseV1 {
    pub id: String,
    pub account_type: String,
    pub owner: String,
    pub address: String,
    pub account_data_hex: String,
    pub expected: DecoderFixtureExpectedV1,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub expected_private_diagnostics: Option<DecoderFixturePrivateDiagnosticsV1>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "kind", rename_all = "camelCase", deny_unknown_fields)]
pub enum DecoderFixtureExpectedV1 {
    Decoded {
        #[serde(rename = "publicValueDigest")]
        public_value_digest: String,
    },
    Error {
        category: DecoderFixtureAccountDecodeErrorCategory,
    },
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum DecoderFixtureAccountDecodeErrorCategory {
    OwnerMismatch,
    UnknownAccountType,
    AccountTypeMismatch,
    AmbiguousAccountType,
    AccountDecodeFailed,
}

impl DecoderFixtureAccountDecodeErrorCategory {
    pub const fn as_str(self) -> &'static str {
        match self {
            Self::OwnerMismatch => "owner_mismatch",
            Self::UnknownAccountType => "unknown_account_type",
            Self::AccountTypeMismatch => "account_type_mismatch",
            Self::AmbiguousAccountType => "ambiguous_account_type",
            Self::AccountDecodeFailed => "account_decode_failed",
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct DecoderFixturePrivateDiagnosticsV1 {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub trailing_bytes: Option<u32>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub candidate_count: Option<u32>,
}

impl DecoderFixtureSetV1 {
    pub fn canonical_projection(&self) -> Result<Self, HashError> {
        validate_decoder_fixture_set_v1(self)?;
        let mut projection = self.clone();
        projection
            .cases
            .sort_by(|left, right| left.id.cmp(&right.id));
        Ok(projection)
    }

    pub fn hash(&self) -> Result<HashId<DecoderFixtureSet>, HashError> {
        hash_decoder_fixture_set_v1(self)
    }
}

pub fn parse_decoder_fixture_set_v1(bytes: &[u8]) -> Result<DecoderFixtureSetV1, HashError> {
    let value = parse_json_bytes_strict(bytes)?;
    let fixture: DecoderFixtureSetV1 =
        serde_json::from_value(value).map_err(|error| projection_error(error.to_string()))?;
    validate_decoder_fixture_set_v1(&fixture)?;
    Ok(fixture)
}

pub fn validate_decoder_fixture_set_v1(fixture: &DecoderFixtureSetV1) -> Result<(), HashError> {
    if fixture.schema != DECODER_FIXTURE_SCHEMA_V1 {
        return Err(HashError::UnknownVersion(fixture.schema.clone()));
    }
    validate_pubkey(&fixture.program_id, "programId")?;
    validate_nonempty_identifier(&fixture.decoder_engine_id, "decoderEngineId", 128)?;
    validate_nonempty_identifier(&fixture.decoder_abi_version, "decoderAbiVersion", 64)?;
    if fixture.cases.is_empty() || fixture.cases.len() > DECODER_FIXTURE_MAX_CASES {
        return invalid(format!(
            "cases must contain between 1 and {DECODER_FIXTURE_MAX_CASES} entries"
        ));
    }

    let mut ids = HashSet::with_capacity(fixture.cases.len());
    let mut total_bytes = 0_usize;
    for case in &fixture.cases {
        validate_stable_id(&case.id, "case id", 128)?;
        if !ids.insert(case.id.as_str()) {
            return invalid(format!("case id '{}' is duplicated", case.id));
        }
        validate_nonempty_identifier(&case.account_type, "accountType", 128)?;
        validate_pubkey(&case.owner, "owner")?;
        validate_pubkey(&case.address, "address")?;
        validate_account_data_hex(&case.account_data_hex)?;
        let account_bytes = case.account_data_hex.len() / 2;
        if account_bytes > DECODER_FIXTURE_MAX_ACCOUNT_BYTES {
            return invalid(format!(
                "case '{}' accountDataHex exceeds {DECODER_FIXTURE_MAX_ACCOUNT_BYTES} bytes",
                case.id
            ));
        }
        total_bytes = total_bytes.checked_add(account_bytes).ok_or_else(|| {
            projection_error("total accountDataHex byte length overflowed".to_string())
        })?;
        if total_bytes > DECODER_FIXTURE_MAX_TOTAL_ACCOUNT_BYTES {
            return invalid(format!(
                "fixture accountDataHex exceeds {DECODER_FIXTURE_MAX_TOTAL_ACCOUNT_BYTES} total bytes"
            ));
        }

        match &case.expected {
            DecoderFixtureExpectedV1::Decoded {
                public_value_digest,
            } => validate_public_value_digest(public_value_digest)?,
            DecoderFixtureExpectedV1::Error { .. } => {}
        }

        if let Some(diagnostics) = &case.expected_private_diagnostics {
            if diagnostics.trailing_bytes.is_none() && diagnostics.candidate_count.is_none() {
                return invalid(format!(
                    "case '{}' expectedPrivateDiagnostics must not be empty",
                    case.id
                ));
            }
            if diagnostics.candidate_count == Some(0) {
                return invalid(format!(
                    "case '{}' candidateCount must be greater than zero",
                    case.id
                ));
            }
        }
    }
    Ok(())
}

pub fn hash_decoder_fixture_set_v1(
    fixture: &DecoderFixtureSetV1,
) -> Result<HashId<DecoderFixtureSet>, HashError> {
    hash_jcs(&fixture.canonical_projection()?)
}

pub fn digest_decoder_fixture_public_value_v1<T: Serialize>(
    value: &T,
) -> Result<String, HashError> {
    let canonical = canonicalize_jcs(value)?;
    Ok(format!(
        "{DECODER_FIXTURE_PUBLIC_VALUE_DIGEST_PREFIX}{}",
        hex::encode(Sha256::digest(canonical))
    ))
}

fn validate_account_data_hex(value: &str) -> Result<(), HashError> {
    if !value.len().is_multiple_of(2)
        || !value
            .as_bytes()
            .iter()
            .all(|byte| byte.is_ascii_digit() || (b'a'..=b'f').contains(byte))
    {
        return invalid("accountDataHex must contain lowercase hexadecimal byte pairs".to_string());
    }
    Ok(())
}

fn validate_public_value_digest(value: &str) -> Result<(), HashError> {
    let Some(digest) = value.strip_prefix(DECODER_FIXTURE_PUBLIC_VALUE_DIGEST_PREFIX) else {
        return invalid("publicValueDigest must use the sha256:<lowercase-hex> format".to_string());
    };
    if digest.len() != 64
        || !digest
            .as_bytes()
            .iter()
            .all(|byte| byte.is_ascii_digit() || (b'a'..=b'f').contains(byte))
    {
        return invalid("publicValueDigest must use the sha256:<lowercase-hex> format".to_string());
    }
    Ok(())
}

fn validate_pubkey(value: &str, field: &str) -> Result<(), HashError> {
    let decoded = bs58::decode(value)
        .into_vec()
        .map_err(|_| projection_error(format!("{field} must be a base58 Solana public key")))?;
    if decoded.len() != 32 || bs58::encode(decoded).into_string() != value {
        return invalid(format!("{field} must be a base58 Solana public key"));
    }
    Ok(())
}

fn validate_nonempty_identifier(
    value: &str,
    field: &str,
    max_length: usize,
) -> Result<(), HashError> {
    if value.is_empty() || value.trim() != value || value.len() > max_length {
        return invalid(format!(
            "{field} must be a nonempty, trimmed string of at most {max_length} bytes"
        ));
    }
    Ok(())
}

fn validate_stable_id(value: &str, field: &str, max_length: usize) -> Result<(), HashError> {
    if value.is_empty()
        || value.len() > max_length
        || !value.as_bytes().iter().enumerate().all(|(index, byte)| {
            byte.is_ascii_lowercase()
                || byte.is_ascii_digit()
                || (index > 0 && matches!(byte, b'-' | b'_' | b'.'))
        })
    {
        return invalid(format!(
            "{field} must be a lowercase stable identifier of at most {max_length} bytes"
        ));
    }
    Ok(())
}

fn projection_error(reason: String) -> HashError {
    HashError::InvalidProjection {
        projection: "decoder fixture set",
        reason,
    }
}

fn invalid<T>(reason: String) -> Result<T, HashError> {
    Err(projection_error(reason))
}

File: typescript/hash/src/fixture.ts
import { sha256 } from "@noble/hashes/sha256";

import { isCanonicalBase58_32 } from "./base58.js";
import { canonicalizeJcs, parseJsonBytesStrict } from "./canonical.js";
import { hashError } from "./error.js";
import { bytesToHex, hashCanonicalPayload } from "./hash.js";
import type {
  DecoderFixtureSetHash,
  IdlNormalizedHash,
  JsonValue,
} from "./types.js";

export const DECODER_FIXTURE_SCHEMA_V2 = "arete.decoder-fixtures/v2" as const;
export const DECODER_FIXTURE_PUBLIC_VALUE_DIGEST_PREFIX = "sha256:" as const;
export const DECODER_FIXTURE_MAX_CASES = 256;
export const DECODER_FIXTURE_MAX_ACCOUNT_BYTES = 1024 * 1024;
export const DECODER_FIXTURE_MAX_TOTAL_ACCOUNT_BYTES = 8 * 1024 * 1024;
export const DECODER_FIXTURE_ACCOUNT_DECODE_ERROR_CATEGORIES = [
  "owner_mismatch",
  "unknown_account_type",
  "account_type_mismatch",
  "ambiguous_account_type",
  "account_decode_failed",
] as const;

export type DecoderFixturePublicValueDigest = `sha256:${string}`;
export type DecoderFixtureAccountDecodeErrorCategory =
  (typeof DECODER_FIXTURE_ACCOUNT_DECODE_ERROR_CATEGORIES)[number];

export interface DecoderFixtureSetV2 {
  readonly schema: typeof DECODER_FIXTURE_SCHEMA_V2;
  readonly programId: string;
  readonly normalizedIdlHash: IdlNormalizedHash;
  readonly decoderEngineId: string;
  readonly decoderAbiVersion: string;
  readonly cases: readonly DecoderFixtureCaseV2[];
}

export interface DecoderFixtureCaseV2 {
  readonly id: string;
  readonly accountType: string;
  readonly owner: string;
  readonly accountDataHex: string;
  readonly expected: DecoderFixtureExpectedV2;
  readonly expectedPrivateDiagnostics?: DecoderFixturePrivateDiagnosticsV2;
}

export type DecoderFixtureExpectedV2 =
  | {
      readonly kind: "decoded";
      readonly publicValueDigest: DecoderFixturePublicValueDigest;
    }
  | {
      readonly kind: "error";
      readonly category: DecoderFixtureAccountDecodeErrorCategory;
    };

export interface DecoderFixturePrivateDiagnosticsV2 {
  readonly trailingBytes?: number;
  readonly candidateCount?: number;
}

export function parseDecoderFixtureSetV2(bytes: Uint8Array): DecoderFixtureSetV2 {
  return validateDecoderFixtureSetV2(parseJsonBytesStrict(bytes));
}

export function validateDecoderFixtureSetV2(value: unknown): DecoderFixtureSetV2 {
  const fixture = expectObject(value, "decoder fixture set");
  expectKeys(
    fixture,
    [
      "schema",
      "programId",
      "normalizedIdlHash",
      "decoderEngineId",
      "decoderAbiVersion",
      "cases",
    ],
    "decoder fixture set",
  );
  if (fixture.schema !== DECODER_FIXTURE_SCHEMA_V2) {
    return hashError(
      "unknown-version",
      `unknown decoder fixture schema '${String(fixture.schema)}'`,
    );
  }
  const programId = expectPubkey(fixture.programId, "programId");
  const normalizedIdlHash = expectNormalizedIdlHash(fixture.normalizedIdlHash);
  const decoderEngineId = expectIdentifier(fixture.decoderEngineId, "decoderEngineId", 128);
  const decoderAbiVersion = expectIdentifier(fixture.decoderAbiVersion, "decoderAbiVersion", 64);
  if (
    !Array.isArray(fixture.cases) ||
    fixture.cases.length === 0 ||
    fixture.cases.length > DECODER_FIXTURE_MAX_CASES
  ) {
    return invalid(`cases must contain between 1 and ${DECODER_FIXTURE_MAX_CASES} entries`);
  }

  const ids = new Set<string>();
  let totalBytes = 0;
  const cases: DecoderFixtureCaseV2[] = fixture.cases.map(
    (caseValue, index): DecoderFixtureCaseV2 => {
      const item = expectObject(caseValue, `cases[${index}]`);
      expectKeys(
        item,
        [
          "id",
          "accountType",
          "owner",
          "accountDataHex",
          "expected",
          "expectedPrivateDiagnostics",
        ],
        `cases[${index}]`,
        ["expectedPrivateDiagnostics"],
      );
      const id = expectStableId(item.id, `cases[${index}].id`, 128);
      if (ids.has(id)) return invalid(`case id '${id}' is duplicated`);
      ids.add(id);
      const accountType = expectIdentifier(
        item.accountType,
        `case '${id}' accountType`,
        128,
      );
      const owner = expectPubkey(item.owner, `case '${id}' owner`);
      const accountDataHex = expectAccountDataHex(item.accountDataHex, id);
      const accountBytes = accountDataHex.length / 2;
      if (accountBytes > DECODER_FIXTURE_MAX_ACCOUNT_BYTES) {
        return invalid(
          `case '${id}' accountDataHex exceeds ${DECODER_FIXTURE_MAX_ACCOUNT_BYTES} bytes`,
        );
      }
      totalBytes += accountBytes;
      if (totalBytes > DECODER_FIXTURE_MAX_TOTAL_ACCOUNT_BYTES) {
        return invalid(
          `fixture accountDataHex exceeds ${DECODER_FIXTURE_MAX_TOTAL_ACCOUNT_BYTES} total bytes`,
        );
      }

      const expected = validateExpected(item.expected, id);
      const expectedPrivateDiagnostics =
        item.expectedPrivateDiagnostics === undefined
          ? undefined
          : validatePrivateDiagnostics(item.expectedPrivateDiagnostics, id);
      return {
        id,
        accountType,
        owner,
        accountDataHex,
        expected,
        ...(expectedPrivateDiagnostics === undefined ? {} : { expectedPrivateDiagnostics }),
      };
    },
  );

  cases.sort((left, right) => compareUtf8(left.id, right.id));
  return {
    schema: DECODER_FIXTURE_SCHEMA_V2,
    programId,
    normalizedIdlHash,
    decoderEngineId,
    decoderAbiVersion,
    cases,
  };
}

export function hashDecoderFixtureSetV2(value: unknown): DecoderFixtureSetHash {
  const projection = validateDecoderFixtureSetV2(value);
  return hashCanonicalPayload(
    "decoder-fixture-set",
    "arete-jcs-v1",
    canonicalizeJcs(projection as unknown as JsonValue),
  );
}

export function digestDecoderFixturePublicValueV2(
  value: JsonValue,
): DecoderFixturePublicValueDigest {
  return `${DECODER_FIXTURE_PUBLIC_VALUE_DIGEST_PREFIX}${bytesToHex(
    sha256(canonicalizeJcs(value)),
  )}`;
}

function validateExpected(value: unknown, id: string): DecoderFixtureExpectedV2 {
  const expected = expectObject(value, `case '${id}' expected`);
  if (expected.kind === "decoded") {
    expectKeys(expected, ["kind", "publicValueDigest"], `case '${id}' expected`);
    return {
      kind: "decoded",
      publicValueDigest: expectPublicValueDigest(expected.publicValueDigest),
    };
  }
  if (expected.kind === "error") {
    expectKeys(expected, ["kind", "category"], `case '${id}' expected`);
    return {
      kind: "error",
      category: expectAccountDecodeErrorCategory(expected.category, id),
    };
  }
  return invalid(`case '${id}' expected.kind must be 'decoded' or 'error'`);
}

function validatePrivateDiagnostics(
  value: unknown,
  id: string,
): DecoderFixturePrivateDiagnosticsV2 {
  const diagnostics = expectObject(value, `case '${id}' expectedPrivateDiagnostics`);
  expectKeys(
    diagnostics,
    ["trailingBytes", "candidateCount"],
    `case '${id}' expectedPrivateDiagnostics`,
    ["trailingBytes", "candidateCount"],
  );
  if (Object.keys(diagnostics).length === 0) {
    return invalid(`case '${id}' expectedPrivateDiagnostics must not be empty`);
  }
  const trailingBytes = diagnostics.trailingBytes === undefined
    ? undefined
    : expectUint32(diagnostics.trailingBytes, `case '${id}' trailingBytes`, true);
  const candidateCount = diagnostics.candidateCount === undefined
    ? undefined
    : expectUint32(diagnostics.candidateCount, `case '${id}' candidateCount`, false);
  return {
    ...(trailingBytes === undefined ? {} : { trailingBytes }),
    ...(candidateCount === undefined ? {} : { candidateCount }),
  };
}

function expectAccountDecodeErrorCategory(
  value: unknown,
  id: string,
): DecoderFixtureAccountDecodeErrorCategory {
  if (
    typeof value !== "string" ||
    !(DECODER_FIXTURE_ACCOUNT_DECODE_ERROR_CATEGORIES as readonly string[]).includes(value)
  ) {
    return invalid(`case '${id}' error category is not a public AccountDecodeErrorCategory`);
  }
  return value as DecoderFixtureAccountDecodeErrorCategory;
}

function expectNormalizedIdlHash(value: unknown): IdlNormalizedHash {
  if (
    typeof value !== "string" ||
    !/^arete:h1:idl-normalized:sha256:[0-9a-f]{64}$/.test(value)
  ) {
    return invalid("normalizedIdlHash must be an idl-normalized typed hash");
  }
  return value as IdlNormalizedHash;
}

function expectPublicValueDigest(value: unknown): DecoderFixturePublicValueDigest {
  if (typeof value !== "string" || !/^sha256:[0-9a-f]{64}$/.test(value)) {
    return invalid("publicValueDigest must use the sha256:<lowercase-hex> format");
  }
  return value as DecoderFixturePublicValueDigest;
}

function expectAccountDataHex(value: unknown, id: string): string {
  if (
    typeof value !== "string" ||
    value.length % 2 !== 0 ||
    !/^[0-9a-f]*$/.test(value)
  ) {
    return invalid(`case '${id}' accountDataHex must contain lowercase hexadecimal byte pairs`);
  }
  return value;
}

function expectPubkey(value: unknown, field: string): string {
  if (!isCanonicalBase58_32(value)) {
    return invalid(`${field} must be a base58 Solana public key`);
  }
  return value;
}

function expectIdentifier(value: unknown, field: string, maxLength: number): string {
  if (
    typeof value !== "string" ||
    value.length === 0 ||
    value.trim() !== value ||
    byteLength(value) > maxLength
  ) {
    return invalid(`${field} must be a nonempty, trimmed string of at most ${maxLength} bytes`);
  }
  return value;
}

function expectStableId(value: unknown, field: string, maxLength: number): string {
  if (
    typeof value !== "string" ||
    byteLength(value) > maxLength ||
    !/^[a-z0-9][a-z0-9._-]*$/.test(value)
  ) {
    return invalid(`${field} must be a lowercase stable identifier of at most ${maxLength} bytes`);
  }
  return value;
}

function expectUint32(value: unknown, field: string, allowZero: boolean): number {
  if (
    !Number.isInteger(value) ||
    (value as number) < (allowZero ? 0 : 1) ||
    (value as number) > 0xffff_ffff
  ) {
    return invalid(`${field} must be ${allowZero ? "a" : "a nonzero"} uint32`);
  }
  return value as number;
}

function expectObject(value: unknown, field: string): Record<string, unknown> {
  if (value === null || typeof value !== "object" || Array.isArray(value)) {
    return invalid(`${field} must be an object`);
  }
  return value as Record<string, unknown>;
}

function expectKeys(
  value: Record<string, unknown>,
  keys: readonly string[],
  field: string,
  optional: readonly string[] = [],
): void {
  const allowed = new Set(keys);
  const optionalKeys = new Set(optional);
  for (const key of Object.keys(value)) {
    if (!allowed.has(key)) invalid(`${field} contains unknown field '${key}'`);
  }
  for (const key of keys) {
    if (!optionalKeys.has(key) && !Object.hasOwn(value, key)) {
      invalid(`${field} is missing '${key}'`);
    }
  }
}

function compareUtf8(left: string, right: string): number {
  const leftBytes = new TextEncoder().encode(left);
  const rightBytes = new TextEncoder().encode(right);
  const length = Math.min(leftBytes.length, rightBytes.length);
  for (let index = 0; index < length; index += 1) {
    if (leftBytes[index] !== rightBytes[index]) return leftBytes[index]! - rightBytes[index]!;
  }
  return leftBytes.length - rightBytes.length;
}

function byteLength(value: string): number {
  return new TextEncoder().encode(value).length;
}

function invalid<T>(reason: string): T {
  return hashError("invalid-projection", `invalid decoder fixture set projection: ${reason}`);
}
